pkg/iam/iam.go

```diff
@@ -15,54 +15,75 @@
 package iam
 
 import (
+	"bytes"
 	"context"
 	"fmt"
+	"strings"
+	"time"
 
 	"google.golang.org/api/cloudresourcemanager/v3"
 
 	"github.com/abcxyz/guardian/pkg/assetinventory"
+	"github.com/abcxyz/guardian/pkg/child"
+	"github.com/abcxyz/pkg/logging"
+	"github.com/sethvargo/go-retry"
 )
 
 // IAM defines the common gcp iam functionality.
 type IAM interface {
 	// OrganizationIAM returns the IAM set on the Organization.
-	OrganizationIAM(ctx context.Context, organizationID string) ([]*AssetIAM, error)
+	OrganizationIAM(ctx context.Context, organizationID string) ([]*assetinventory.AssetIAM, error)
 	// FolderIAM returns the IAM set on the Folder.
-	FolderIAM(ctx context.Context, folderID string) ([]*AssetIAM, error)
+	FolderIAM(ctx context.Context, folderID string) ([]*assetinventory.AssetIAM, error)
 	// ProjectIAM returns the IAM set on the Project.
-	ProjectIAM(ctx context.Context, projectID string) ([]*AssetIAM, error)
-}
-
-// AssetIAM represents the IAM of a GCP resource (e.g binding/policy/membership of GCP Project, Folder, Org).
-type AssetIAM struct {
-	// The ID of the resource (e.g. Project ID, Folder ID, Org ID).
-	ResourceID string
-	// The type of the resource (e.g. Project, Folder, Org).
-	ResourceType string
-	// The IAM membership (e.g. group:my-group@google.com).
-	Member string
-	// The role (e.g. roles/owner).
-	Role string
+	ProjectIAM(ctx context.Context, projectID string) ([]*assetinventory.AssetIAM, error)
+	// RemoveOrganizationIAM removes the given IAM policy membership.
+	RemoveOrganizationIAM(ctx context.Context, projectIAMMember *assetinventory.AssetIAM) error
+	// RemoveFolderIAM removes the given IAM policy membership.
+	RemoveFolderIAM(ctx context.Context, projectIAMMember *assetinventory.AssetIAM) error
+	// RemoveProjectIAM removes the given IAM policy membership.
+	RemoveProjectIAM(ctx context.Context, projectIAMMember *assetinventory.AssetIAM) error
 }
 
 type IAMClient struct {
 	crmService *cloudresourcemanager.Service
+	cfg        *Config
+	workingDir *string
 }
 
+// Config is the config values for the IAM client.
+type Config struct {
+	maxRetries        uint64
+	initialRetryDelay time.Duration
+	maxRetryDelay     time.Duration
+}
+
+const (
+	concurrentPolicyErrMessage = "is the subject of a conflict: There were concurrent policy changes"
+)
+
 // NewClient creates a new iam client.
-func NewClient(ctx context.Context) (*IAMClient, error) {
+func NewClient(ctx context.Context, workingDir *string) (*IAMClient, error) {
 	crm, err := cloudresourcemanager.NewService(ctx)
 	if err != nil {
 		return nil, fmt.Errorf("failed to initialize cloudresourcemanager service: %w", err)
 	}
 
+	cfg := &Config{
```

Comment:
As a follow up, i created a reusable retry config that you can inject to each command, see here: https://github.com/abcxyz/guardian/blob/main/pkg/commands/plan/plan_init.go#L51

Reply:
Ack